client/src/lib.rs

```diff
@@ -114,31 +113,30 @@ async fn execute_task<'a, T, E: From<ClientError>>(
     alloc: &ResourceAlloc,
 ) -> Result<T, E> {
     task.init(Some(alloc))?;
-    let mut cont = TaskResult::Continue;
-    while cont == TaskResult::Continue {
-        let mut wait = true;
-        while wait {
-            let preemtive_state = wait_preemptive(client, timeout).await?;
-            match preemtive_state {
-                PreemptionResponse::Wait => tokio::time::sleep(Duration::from_millis(1000)).await,
-                PreemptionResponse::Execute => {
-                    //final check client-side execution?
-                    wait = false;
-                }
-                PreemptionResponse::Abort => {
-                    warn!("Client {} aborted", client.token.process_id());
-                    release_preemptive(client).await?;
-                    return task.end(Some(alloc));
+    loop {
+        let preemtive_state = wait_preemptive(client, timeout).await?;
+        match preemtive_state {
+            PreemptionResponse::Wait => {
+                tokio::time::sleep(Duration::from_millis(1000)).await;
+                continue;
+            }
+            PreemptionResponse::Execute => {
+                let cont = task.task(Some(alloc))?;
+                release_preemptive(client).await?;
+                info!(
+                    "Client {} task iteration completed",
+                    client.token.process_id()
+                );
+                if cont == TaskResult::Done {
+                    break;
                 }
             }
+            PreemptionResponse::Abort => {
+                warn!("Client {} aborted", client.token.process_id());
+                release_preemptive(client).await?;
+                return Err(E::from(ClientError::Aborted));
```

Comment:
Shouldn't there be a release(client) here? release_preemtive does not release memory.